xla/service/cpu/onednn_ops_rewriter.cc

```diff
@@ -143,183 +139,288 @@ std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
     return std::nullopt;
   }
 
-  if (left_producer != right_producer || left_producer->user_count() != 2)
+  if (left_producer != right_producer || left_producer->user_count() != 2) {
     return std::nullopt;
+  }
 
   return left_producer;
 }
 
+auto MeanPattern(HloInstruction** input) {
+  return m::Reshape(
+      m::Convert(m::Divide(m::Reduce(m::Convert(m::Op(input)), m::Op()),
+                           m::Broadcast(m::Convert()))));
+}
+
+template <typename Pattern>
+auto Square(Pattern pattern) {
+  return m::Multiply()
+      .WithBinaryOperandsAnyOrder(pattern, pattern)
+      .WithPredicate([](const HloInstruction* instr) {
+        return instr->unique_operands().size() == 1;
+      });
+}
+
+std::optional<bool> MatchTFKerasLayerNorm(HloInstruction* instr,
+                                          HloInstruction** src,
+                                          HloInstruction** scale,
+                                          HloInstruction** bias, float* eps) {
+  // variance = Mean((X - Mean(x))^2)
+  // Z = scale / sqrt(variance + eps)
+  // LN(X) = X*Z + Bias - Mean(X)*Z
+
+  HloInstruction *src_a, *src_b, *src_c;
+  HloInstruction *bias_node, *scaled_norm_a, *scaled_norm_b, *mean0_a, *epsilon,
+      *sqrd_diff_mean, *scale_node, *sqrd_diff;
+
+  // First Match X*Z + Bias - Mean(X)*Z
+
+  if (!Match(
+          instr,
+          m::Add().WithBinaryOperandsAnyOrder(
+              m::Multiply()
+                  .WithBinaryOperandsAnyOrder(m::Op(src), m::Op(&scaled_norm_a))
+                  .WithOneUser(),
+              m::Subtract(m::Op(&bias_node),
+                          m::Multiply().WithBinaryOperandsAnyOrder(
+                              m::Broadcast(m::Reshape(m::Op(&mean0_a))),
+                              m::Op(&scaled_norm_b)))
+                  .WithOneUser()))) {
+    return std::nullopt;
+  }
+
+  if (scaled_norm_a != scaled_norm_b) return std::nullopt;
+
+  const Shape& src_shape = (*src)->shape();
+  if (!IsSupportedType(src_shape.element_type())) return std::nullopt;
+
+  // get bias
```

Comment:
Nit: Could you please capitalize the first letter of each comment?
```suggestion
  // Get bias
```

Reply:
Done